Firestore/core/src/firebase/firestore/immutable/array_sorted_map.h

```diff
@@ -0,0 +1,306 @@
+  const_iterator begin() const {
+    return contents_.begin();
+  }
+
+  iterator end() {
+    return begin() + size_;
+  }
+
+  const_iterator end() const {
+    return begin() + size_;
+  }
+
+ private:
+  array_type contents_;
+  size_type size_;
+
+  template <typename K, typename V, typename C>
+  friend class ArraySortedMap;
+};
+
+/**
+ * ArraySortedMap is a value type containing a map. It is immutable, but has
+ * methods to efficiently create new maps that are mutations of it.
+ */
+template <typename K, typename V, typename C = std::less<K>>
+class ArraySortedMap : public ArraySortedMapBase {
+ public:
+  typedef ArraySortedMap<K, V, C> this_type;
+  typedef KeyComparator<K, V, C> key_comparator_type;
+
+  /**
+   * The type of the entries stored in the map.
+   */
+  typedef std::pair<K, V> value_type;
+
+  /**
+   * The type of the fixed-size array containing entries of value_type.
+   */
+  typedef FixedArray<value_type, kFixedSize> array_type;
+  typedef typename array_type::const_iterator const_iterator;
+
+  typedef std::shared_ptr<const array_type> array_pointer;
+
+  /**
+   * Creates an empty ArraySortedMap.
+   */
+  explicit ArraySortedMap(const C& comparator = C())
+      : array_(EmptyArray()), key_comparator_(comparator) {
+  }
+
+  /**
+   * Creates an ArraySortedMap containing the given entries.
+   */
+  explicit ArraySortedMap(std::initializer_list<value_type> entries,
+                          const C& comparator = C())
+      : array_(), key_comparator_(comparator) {
+    assert(static_cast<size_type>(entries.size()) <= kFixedSize);
+
+    auto array = std::make_shared<array_type>();
+    array->append(entries.begin(), entries.end());
+    array_ = array;
+  }
+
+  /**
+   * Creates a new map identical to this one, but with a key-value pair added or
+   * updated.
+   *
+   * @param key The key to insert/update.
+   * @param value The value to associate with the key.
+   * @return A new dictionary with the added/updated value.
+   */
+  this_type insert(const K& key, const V& value) const {
+    value_type pair(key, value);
+
+    const_iterator current_end = end();
+    const_iterator pos = LowerBound(pair.first);
+    bool replacing_entry = false;
+
+    if (pos != current_end) {
+      // LowerBound found an entry where pos->first >= pair.first. Reversing the
+      // argument order here tests pair.first < pos->first.
+      bool keys_equal = !key_comparator_(pair.first, *pos);
+      if (keys_equal) {
+        const V& pos_value = pos->second;
+        const V& pair_value = pair.second;
+        if (pos_value == pair_value) {
+          return *this;
+        } else {
+          replacing_entry = true;
+        }
+      }
+    }
+
+    // Copy the segment before the found position. If not found, this is
+    // everything.
+    auto copy = std::make_shared<array_type>();
+    copy->append(begin(), pos);
+
+    // Copy the value to be inserted.
+    copy->append(std::move(pair));
+
+    if (replacing_entry) {
```

Comment:
Optional nit: consider making 'pos' const. Then this could be:

```
if (replacing_entry) {
  copy->append(pos + 1, current_end);
} else {
  assert(size() < kFixedSize);
  copy->append(pos, current_end);
}
```

I find this slightly clearer, but it's a matter of taste.